include/seqan3/alignment/matrix/detail/alignment_trace_matrix_full_banded.hpp

```diff
@@ -0,0 +1,238 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::alignment_trace_matrix_full_banded.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/alignment/band/static_band.hpp>
+#include <seqan3/alignment/matrix/detail/alignment_matrix_column_major_range.hpp>
+#include <seqan3/alignment/matrix/detail/alignment_trace_matrix_base.hpp>
+#include <seqan3/alignment/matrix/detail/alignment_trace_matrix_proxy_type.hpp>
+#include <seqan3/std/iterator>
+#include <seqan3/std/ranges>
+
+namespace seqan3::detail
+{
+
+/*!\brief An alignment traceback matrix storing the entire banded traceback matrix.
+ * \tparam trace_t          The type of the trace directions.
+ * \tparam coordinate_only  A boolean flag indicating if only a seqan3::alignment_coordinate should be generated.
+ * \ingroup alignment_matrix
+ *
+ * \details
+ *
+ * This implementation allocates the full banded traceback matrix using `k*m` memory, where `k` is the band size and
+ * `m` the number of columns to store. The matrix allows access to the
+ * underlying values through a range based interface. An iterator over the traceback matrix iterates in
+ * column-major-order over the traceback matrix. Dereferencing an iterator returns a slice over the current matrix
+ * column. The value type is a pair over a seqan3::alignment_coordinate and
+ * the seqan3::detail::alignment_trace_matrix_proxy_type, which gives a unified access to the respective matrix cells
+ * as needed by the standard alignment algorithm. The matrix is modelled as std::ranges::InputRange since the
+ * alignment algorithm iterates only once over the complete matrix to calculate the values.
+ *
+ * ### Only computing the coordinates
+ *
+ * Sometimes it is desired to only get access to the alignment coordinates. This can be achieved by setting
+ * `coordinate_only = true`. In this case no memory will be allocated and only an internal state is maintained to
+ * generate the alignment coordinates. The coordinates are relative to the banded matrix, i.e. they do not represent the
+ * corresponding row indices of the actual (unbanded) matrix.
+ */
+template <typename trace_t, bool coordinate_only = false>
+class alignment_trace_matrix_full_banded :
```

Comment:
I really don't think that this will help making the design more clear. A lot of stuff is already outsourced. I'll see that I can remove the definition of the noop functions by putting defaults into the base class. But the most important point is that I don't want to define variables (especially in SIMD mode) if the algorithm don't use them.